src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -380,96 +509,60 @@ public void Stop()
-                            flushResult = await _frameWriter.WriteDataAndTrailersAsync(_stream, _flowControl, readResult.Buffer, firstWrite, _stream.ResponseTrailers);
-                        }
-                        else
-                        {
-                            flushResult = await _frameWriter.WriteResponseTrailersAsync(_stream, _stream.ResponseTrailers);
-                        }
-                    }
-                    else if (readResult.IsCompleted && _streamEnded)
-                    {
-                        if (readResult.Buffer.Length != 0)
-                        {
-                            ThrowUnexpectedState();
-                        }
+                // Noop, we're done
+                return;
+            }
 
-                        // Headers have already been written and there is no other content to write
-                        flushResult = await _frameWriter.FlushAsync(outputAborter: null, cancellationToken: default);
-                    }
-                    else
-                    {
-                        var endStream = readResult.IsCompleted;
-                        flushResult = await _frameWriter.WriteDataAsync(_stream, _flowControl, readResult.Buffer, endStream, firstWrite, forceFlush: true);
-                    }
+            _requestProcessingComplete = true;
 
-                    firstWrite = false;
-                    _pipeReader.AdvanceTo(readResult.Buffer.End);
-                } while (!readResult.IsCompleted);
-            }
-            catch (Exception ex)
+            if (_completedResponse)
             {
-                _log.LogCritical(ex, nameof(Http2OutputProducer) + "." + nameof(ProcessDataWrites) + " observed an unexpected exception.");
+                Stream.CompleteStream(errored: false);
             }
+        }
+    }
 
-            await _pipeReader.CompleteAsync();
+    internal void CompleteResponse()
+    {
+        lock (_dataWriterLock)
+        {
+            if (_completedResponse)
+            {
+                // This should never be called twice
+                return;
+            }
 
-            // Signal via WriteStreamSuffixAsync to the stream that output has finished.
-            // Stream state will move to RequestProcessingStatus.ResponseCompleted
-            _responseCompleteTaskSource.SetResult(flushResult);
+            _completedResponse = true;
 
-            // Wait here for the stream to be reset or disposed.
-            await new ValueTask(_resetAwaitable, _resetAwaitable.Version);
-            _resetAwaitable.Reset();
-        } while (!_disposed);
+            if (_resetErrorCode is { } error)
+            {
+                // If we have an error code to write, write it now that we're done with the response.
+                // Always send the reset even if the response body is completed. The request body may not have completed yet.
+                _ = _frameWriter.WriteRstStreamAsync(StreamId, error).Preserve();
```

Comment:
Can we make CompleteResponse() Task-returning and await it in WriteToOutputPipe?

Reply:
Done